winter/converters/converter.py

```diff
@@ -0,0 +1,153 @@
+import datetime
+import enum
+import inspect
+import typing
+
+import dataclasses
+from dateutil import parser
+
+from ..type_utils import get_origin_type
+from ..type_utils import is_optional
+
+_converters = {}
+
+
+class ConvertError(Exception):
+
+    def __init__(self, errors):
+        self.errors = errors
+
+
+def converter(type_: typing.Type, validator: typing.Callable = None):
+    def wrapper(func):
+        converters = _converters.setdefault(type_, [])
+        converters.append((func, validator))
+        return func
+
+    return wrapper
+
+
+def convert(data, hint_class):
+    origin_type = get_origin_type(hint_class)
+
+    types_ = origin_type.mro() if inspect.isclass(origin_type) else type(origin_type).mro()
+
+    for type_ in types_:
+        converters = _converters.get(type_, [])
+
+        for converter_, checker in converters:
+            if checker is None or checker(hint_class):
+                return converter_(data, hint_class)
+
+    raise ConvertError(f'Cannot convert "{data}"')
+
+
+@converter(object, validator=is_optional)
+def optional_converter(data, type_):
+    if data is None:
+        return None
+    type_ = type_.__args__[0]
+    return convert(data, type_)
+
+
+@converter(object, validator=dataclasses.is_dataclass)
+def dataclass_converter(data: typing.Dict[str, typing.Any], type_):
+    if not isinstance(data, dict):
+        raise ConvertError(f'Cannot convert "{data}"')
+
+    errors = {}
+    converted_data = {}
+    missing_fields = []
+
+    for field in dataclasses.fields(type_):
+        field_data = data.get(field.name, dataclasses.MISSING)
+
+        if field_data is dataclasses.MISSING:
+            if is_optional(field.type):
+                field_data = None
+            else:
+                missing_fields.append(field.name)
+                continue
+        try:
+            field_data = convert(field_data, field.type)
+        except ConvertError as e:
+            errors[field.name] = e.errors
+        else:
+            converted_data[field.name] = field_data
+    if missing_fields:
+        missing_fields = '", "'.join(missing_fields)
+        errors['non_field_error'] = f'Missing fields: "{missing_fields}"'
```

Comment:
Usually, it's a good idea to keep exceptions structured instead of formatted text.